app/lib/boolean-phrase-parser.js

```diff
@@ -0,0 +1,208 @@
+                    binaryOperator = currentChar;
+                    leftOperatorEndIndex = i;
+                }
+
+            } else if (closeParentheses > openParentheses) {
+                //TODO: add some more information here
+                throw new Error('mismatch of parentheses');
+            }
+        }
+
+        if (binaryOperator !== null) {
+            var leftOperand = expression.substring(0, leftOperatorEndIndex).trim();
+            var rightOperand =  expression.substring(leftOperatorEndIndex + 1, expression.length + 1).trim();
+            if(!(leftOperand && rightOperand)){
+                throw new Error('one or more operands are missing in expression: ' +  expression);
+            }
+            // exp & exp
+            return {
+                leftOperand: leftOperand,
+                binaryOperator: binaryOperator,
+                rightOperand: rightOperand
+            };
+        } else { //binaryOperator === null
+            if (expression[0] === openParethesisChar) {
+                if (openParentheses === closeParentheses) {
+                    if(expression[expression.length - 1] === closeParethesisChar) {
+                        // (exp)
+                        return {
+                            parenthesesExpression: true
+                        };
+                    } else {
+                        // (exp)bla
+                        throw new Error('operator is missing between two operands');
+                    }
+                } else if (openParentheses > closeParentheses) {
+                    //((exp) unclosed parentheses case (the other case we should be getting on the previous code
+                    //TODO add some more info here
+                    throw new Error('unclosed parentheses');
+                } else {
+                    // (exp)) this case is actually handled when running through the chars (just in case)
+                    throw new Error('mismatch of parentheses');
+                }
+            } else if (expression[0] === notChar) {
+                //!exp
+                return {
+                    unaryOperator: notChar,
+                    expression: expression.substring(1, expression.length + 1)
+                };
+            } else if (openParentheses === 0) {
+                //literal
+                return {
+                    literal: expression
+                };
+            } else {
+                //there must be an error here
+                throw new Error('should not have happened. expresion does not fall into any one of the patterns');
+            }
+        }
+    };
+
+    ExpressionNode.prototype.isBinaryOperand = function (char) {
+        return char === orChar || char === andChar;
+    };
+
+    ExpressionNode.prototype.isWhiteSpaceChar = function (char) {
+        return char === ' ' || char === '\t' || char === '\n' || char === '\r';
```

Comment:
I think this should be something like: return char !== char.trim(); to cover all the cases